### PR TITLE
misc fixes

### PR DESCRIPTION
**Pull request by zoulasc**
* Add a function to copy a string with a string with a larger allocation
   (to be used by the case folding routines)
* Add printf attributes to the printf-like functions and fix one format warning
* Cleanup the tempfree macro
* make more functions static
* rename fp to frp (FRame Pointer) to avoid shadowing with fp (File Pointer).
* add more const
* fix indent in UPLUS case
* add locale-aware case folding
* make nfiles size_t
* fix bugs in file closing:
        - compare fclose to EOF and pclose to -1
        - use nfiles instead of FOPEN_MAX in closeall
        - don't close files we did not open (0,1,2) fpurge/fflush instead
* Add a test for german case folding.

### CODE REVIEW

File: testdir/T.builtin [hunk 1]
```diff
@@ -29,6 +29,12 @@ $awk '{ printf("%s|%s|%s\n", tolower($0), toupper($0), $0)}' >foo1
 echo 'hello, world!|HELLO, WORLD!|hello, WORLD!' >foo2
 diff foo1 foo2 || echo 'BAD: T.builtin (toupper/tolower)'
 
+
+(export LANG=de_DE.UTF-8 && echo 'Dürst' |
+$awk '{ printf("%s|%s|%s\n", tolower($0), toupper($0), $0)}') >foo1
+echo 'dürst|DÜRST|Dürst' >foo2
+diff foo1 foo2 || echo 'BAD: T.builtin (toupper/tolower) for utf-8'
```
Comment (arnoldrobbins) [inline testdir/T.builtin]: Please wrap this is in something like:
```sh
if locale -a | grep -i de_DE.UTF-8 > /dev/null
then
     ...
fi ```

File: proto.h [hunk 2]
```diff
@@ -135,9 +136,12 @@ extern	void	yyerror(const char *);
...
 extern	void	bcheck2(int, int, int);
-extern	void	SYNTAX(const char *, ...);
-extern	void	FATAL(const char *, ...) __attribute__((__noreturn__));
-extern	void	WARNING(const char *, ...);
+extern	void	SYNTAX(const char *, ...)
+    __attribute__((__format__(__printf__, 1, 2)));
+extern	void	FATAL(const char *, ...)
+    __attribute__((__format__(__printf__, 1, 2), __noreturn__));
+extern	void	WARNING(const char *, ...)
+    __attribute__((__format__(__printf__, 1, 2)));
```
[the remark this answers is not in the document]
Comment (zoulasc) [reply proto.h]: __attribute__ is already used in proto.h:
extern  void    overflo(const char *) __attribute__((__noreturn__));

Should I add ifdefs to make it empty if not gcc, like other projects do?

File: run.c [hunk 3]
```diff
@@ -1505,16 +1507,73 @@ Cell *instat(Node **a, int n)	/* for (a[0] in a[1]) a[2] */
...
+			ps += n;
+
+			n = wcrtomb(pbuf, fun_wc(wc), &mbs2);
+			if (n == (size_t)-1)
+				FATAL("illegal wide character %s", s);
+
+			pbuf += n;
+		}
+
+		*pbuf = 0;
```
[the remark this answers is not in the document]
Comment (zoulasc) [reply run.c]: To be consistent would require a larger patch. Should I do it?
http://www.netbsd.org/~christos/run.c.diff

File: run.c [hunk 4]
```diff
@@ -1765,52 +1819,59 @@ const char *filename(FILE *fp)
...
+			FATAL("i/o error occurred closing %s", files[i].fname);
+		if (i > 2)	/* don't do /dev/std... */
+			xfree(files[i].fname);
+		files[i].fname = NULL;	/* watch out for ref thru this */
+		files[i].fp = NULL;
 	}
 	tempfree(x);
 	x = gettemp();
-	setfval(x, (Awkfloat) stat);
+	setfval(x, (Awkfloat) stat ? -1 : 0);
```
[the remark this answers is not in the document]
Comment (zoulasc) [reply run.c]: good catch.

File: testdir/T.builtin (hunk 1 above)
Comment (zoulasc) [reply testdir/T.builtin]: ok. "fgrep -qsi ..."

File: proto.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (arnoldrobbins) [reply proto.h]: Harumph. I didn't know it was already used. Let me think about this a little longer.

File: run.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (arnoldrobbins) [reply run.c]: Yes, being consistent on `'\0'` would be good. Thanks.

File: testdir/T.builtin (hunk 1 above)
Comment (arnoldrobbins) [reply testdir/T.builtin]: OK.

File: proto.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (arnoldrobbins) [reply proto.h]: You can add these changes.  Please update the whole PR.  Thanks.

File: run.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (arnoldrobbins) [reply run.c]: Looks like you didn't fix this one yet. Please do.

**Pull request merged**